poly/promise/promise.go

```diff
@@ -0,0 +1,744 @@
+	p.numShares = 0
+	p.priShares = new(poly.PriShares)
+	p.priShares.Empty(p.shareSuite, p.t, p.n)
+	
+	// Populate the secrets array. It encrypts each share with a diffie
+	// hellman exchange between the originator of the promist and the
+	// specific guardian.
+	for i := 0 ; i < p.n; i++ {
+		diffieBase := p.shareSuite.Point().Mul(guardians[i], keyPair.Secret)
+		p.secrets[i] = p.diffieHellmanEncrypt(prishares.Share(i), diffieBase)
+	}
+	
+	return p
+}
+
+// Returns the id of the policy
+func (p *Promise) GetId() string {
+	return p.id
+}
+
+/* Given a Diffie-Hellman shared key, encrypts a secret.
+ *
+ * Arguments
+ *    secret      = the secret to encrypt
+ *    diffieBase  = the DH shared key
+ *
+ * Return
+ *   the encrypted secret
+ */
+func (p *Promise) diffieHellmanEncrypt(secret abstract.Secret, diffieBase abstract.Point) abstract.Secret {	
+	buff, err := diffieBase.MarshalBinary()
+	if err != nil {
+		panic("Bad shared secret for Diffie-Hellman give.")
+	}
+	cipher := p.shareSuite.Cipher(buff)
+	diffieSecret := p.shareSuite.Secret().Pick(cipher)
+	return p.shareSuite.Secret().Add(secret, diffieSecret)
+}
+
+/* Given a Diffie-Hellman shared key, decrypts a secret.
+ *
+ * Arguments
+ *    secret      = the secret to decrypt
+ *    diffieBase  = the DH shared key
+ *
+ * Return
+ *   the decrypted secret
+ */
+func (p *Promise) diffieHellmanDecrypt(secret abstract.Secret, diffieBase abstract.Point) abstract.Secret {	
+	buff, err := diffieBase.MarshalBinary()
+	if err != nil {
+		panic("Bad shared secret for Diffie-Hellman give.")
+	}
+	cipher := p.shareSuite.Cipher(buff)
+	diffieSecret := p.shareSuite.Secret().Pick(cipher)
+	return p.shareSuite.Secret().Sub(secret, diffieSecret)
+}
+
+/* This helper function makes sure that a share is "syntactically" valid. In
+ * other words, the index is properly in range and the guardian it was sent to
+ * the correct guardian. VerifyShare handles "semantic" verification.
+ *
+ * Arguments
+ *   i        = the index of the share
+ *   gKeyPair = the key pair of the guardian of the share
+ *
+ * Returns
+ *    Whether the share is valid or invalid.
+ */
+func (p *Promise) validShare(i int, gKeyPair *config.KeyPair) bool {
+	if i < 0 || i > p.n {
+		return false
+	}
+	if !p.guardians[i].Equal(gKeyPair.Public) {
+		return false
+	}
+	return true
+}
+
+
+/* Verify that a share has been properly constructed.
+ *
+ * Arguments
+ *    i        = the index of the share to verify
+ *    gPrikey  = the private key of the guardian of share i
+ *
+ * Return
+ *   whether the decrypted secret properly passes the public polynomial.
+ *
+ * Note
+ *   Make sure that the proper index and key is specified. Otherwise, the
+ *   function will return false because diffieHellmanDecrypt gave the wrong
+ *   result. In short, make sure to verify only shares that are allotted to you.
+ */
+func (p *Promise) VerifyShare(i int, gKeyPair *config.KeyPair) bool {
+	if !p.validShare(i, gKeyPair) {
```

Comment:
This was primarily an artifact of an initial design. I was planning on merging the two and did so in a recent commit.